### PR TITLE
Software

### PR DESCRIPTION
**Pull request by King-Of-Nothing25**

### CODE REVIEW

File: software/main.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    try:
+        while True:
+            #referee handling
+            if competition:
+                cmd = robotReferee.getCommand()
+                if (cmd != None):
+                    if cmd[0] == 'START':
+                        stateMachine.currentState = State.FIND_BALL
+                        if cmd[1] == 'blue':
+                            stateMachine.throwIntoBlue = True
```
Comment (Akustav) [inline software/main.py:53]: Use color enum instead of a boolean

File: software/motion.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        #calculate speeds to send over serial
+        for i in range(3):
+            wheelAngularSpeedInMainboardUnits[i] = wheelLinearVelocities[i] * self.wheelSpeedToMainboardUnits
+
+        self.serialCommunication(int(wheelAngularSpeedInMainboardUnits[0]), int(wheelAngularSpeedInMainboardUnits[1]), int(wheelAngularSpeedInMainboardUnits[2]), int(throwerSpeed))
+
+    #orbit movement around point to find basket
+    #positive orbit speed is orbiting orbiting counterclockwise if viewed from above
+    def orbit(self, orbitSpeed, orbitRadius):
+        orbitCircumference = orbitRadius * 2 * 3.141528
```
Comment (Akustav) [inline software/motion.py:56]: Why not use math.pi

File: software/referee.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
+import websocket
+import time
+import json
+import multiprocessing
+
+class Referee:
+    def __init__(self, ip, robotName = "snax"):
+        self.robotName = robotName
+        self.serverURL = "ws://" + ip + ":8111"
```
Comment (Akustav) [inline software/referee.py:9]: Maybe have the port configurable as well

File: software/statemachine.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+            if self.imageData.basket_m.exists:
+                trajectorySpeed = 0.2
+                if self.imageData.basket_m.x > (self.imageWidth / 2):
+                    trajectorySpeed = -trajectorySpeed
+                    rotSpeed = -rotSpeed
+
+        self.robot.move(trajectorySpeed, 0, rotSpeed)
+
+    #averages pixel distances of a 3x3 square from depth_image
+    def basketDist(self, x, y):
```
Comment (Akustav) [inline software/statemachine.py:177]: calculate distance in image processor. There is a place there.

**Review by Akustav — COMMENTED** [review]
Minor issues

File: software/statemachine.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+                if self.imageData.basket_m.x > (self.imageCenterPoint):
+                    trajectorySpeed = -trajectorySpeed
+                    rotSpeed = -rotSpeed
+
+        self.robot.move(trajectorySpeed, 0, rotSpeed)
+
+    #averages pixel distances of a 5x5 square from depth_image
+    def basketDist(self, x, y):
+        distanceSum = 0
+        for i in range(-2, 3):
```
Comment (Akustav) [inline software/statemachine.py:192]: You can select a range of pixels using numpy and use the np.average method for that

File: software/image_processor.py [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+                        y -= 1
+                    if direction == "y":
+                        y += 1
+                    if direction == "-x":
+                        x -= 1
+                    if direction == "x":
+                        x += 1
+                return x, y
+            except:
+                return False
```
Comment (Akustav) [inline software/image_processor.py:93]: You currently return different data based on a methods exit conditions. This is error prone

File: software/image_processor.py [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+    def set_segmentation_table(self, table):
+        segment.set_table(table)
+
+    def start(self):
+        self.camera.open()
+
+    def stop(self):
+        self.camera.close()
+
+    def find_multiple_matching_pixels(self,color, x, y, image, direction, pixels):
```
Comment (Akustav) [inline software/image_processor.py:75]: What does this method do?

**Review by Akustav — COMMENTED** [review]
Comments

File: software/main.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+            
+            
+            # has argument aligned_depth that enables depth frame to color frame alignment. Costs performance
+            processedData = processor.process_frame(aligned_depth=useDepthImage)
+            #updating state machine data
+            stateMachine.setImageData(processedData)
+
+            useDepthImage = False
+
+            if stateMachine.currentState == statemachine.State.FIND_BALL:
```
Comment (Akustav) [inline software/main.py:71]: Why have this if block? It sets the state state to the value of the current state? Maybe set the depth boolean if the state is one of the two that currently trigger it.

File: software/statemachine.py [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+    GO_TO_BASKET = 7
+
+class StateMachine:
+    def __init__(self, omniRobot):
+        self.currentState = State.FIND_BALL
+        self.robot = omniRobot
+        self.throwIntoBlue = False
+
+        self.imageData = None
+        self.imageWidth = 848
```
Comment (Akustav) [inline software/statemachine.py:20]: Should the camera not define this? Else if you change it, this part of the logic will break.

**Review by Akustav — COMMENTED** [review]
Needs a little work

File: software/statemachine.py (hunk 5 above)
Comment (dr3san) [reply software/statemachine.py:211]: Good to know

File: software/image_processor.py (hunk 7 above)
Comment (dr3san) [reply software/image_processor.py:75]: @King-Of-Nothing25

File: software/image_processor.py (hunk 6 above)
Comment (dr3san) [reply software/image_processor.py:93]: @King-Of-Nothing25

File: software/statemachine.py (hunk 4 above)
Comment (dr3san) [reply software/statemachine.py:209]: Added basket distance calculation in image_processor.py analyze_baskets(). Statemachine now gets the distance as a field of the basket object.

File: software/statemachine.py (hunk 9 above)
Comment (dr3san) [reply software/statemachine.py:20]: Removed hardcoded values.

File: software/main.py (hunk 8 above)
Comment (dr3san) [reply software/main.py:71]: The setState() method was written when we didn't understand that the statemachine can't use blocking methods. Replaced with executeState() method. Main loop is much cleaner now

File: software/image_processor.py (hunk 6 above)
Comment (dr3san) [reply software/image_processor.py]: I removed this code, we didn't use it

**Review by dr3san — APPROVED** [review]
approve changes...?

**Pull request merged**